templates/commands/render/templatesource/git.go

```diff
@@ -0,0 +1,218 @@
+	case "ssh":
+		remote = string(g.re.ExpandString(nil, g.sshRemoteExpansion, src, match))
+	default:
+		return nil, false, fmt.Errorf("protocol %q isn't usable with a template sourced from a remote git repo", protocol)
+	}
+
+	return &gitDownloader{
+		remote:  remote,
+		subdir:  string(g.re.ExpandString(nil, g.subdirExpansion, src, match)),
+		version: string(g.re.ExpandString(nil, g.versionExpansion, src, match)),
+		cloner:  &realCloner{},
+		tagser:  &realTagser{},
+	}, true, nil
+}
+
+// gitDownloader implements templateSource for templates hosted in a remote git
+// repo, regardless of which git hosting service it uses.
+type gitDownloader struct {
+	// An HTTPS or SSH connection string understood by "git clone"
+	remote string
+	// An optional subdirectory within the git repo that we want
+	subdir string
+
+	// either a semver-formatted tag, or the magic value "latest"
+	version string
+
+	cloner cloner
+	tagser tagser
+}
+
+// Download implements templateDownloader.
+func (g *gitDownloader) Download(ctx context.Context, outDir string) error {
+	logger := logging.FromContext(ctx).With("logger", "gitDownloader.Download")
+
+	// Validate first before doing expensive work
+	subdir, err := common.SafeRelPath(nil, g.subdir) // protect against ".." traversal attacks
+	if err != nil {
+		return fmt.Errorf("invalid subdirectory: %w", err)
+	}
+
+	branchOrTag, err := resolveBranchOrTag(ctx, g.tagser, g.remote, g.version)
+	if err != nil {
+		return err
+	}
+	logger.DebugContext(ctx, "resolved version to branchOrTag", "version", g.version, "branchOrTag", branchOrTag)
+
+	// Rather than cloning directly into outDir, we clone into a temp dir. It would
+	// be incorrect to clone the whole repo into outDir if the caller only asked
+	// for a subdirectory, e.g. "github.com/my-org/my-repo/my-subdir@v1.2.3".
+	tmpDir, err := os.MkdirTemp(os.TempDir(), "git-clone-")
+	if err != nil {
+		return fmt.Errorf("MkdirTemp: %w", err)
+	}
+	defer os.RemoveAll(tmpDir)
+	subdirToCopy := filepath.Join(tmpDir, filepath.FromSlash(subdir))
+
+	if err := g.cloner.Clone(ctx, g.remote, branchOrTag, tmpDir); err != nil {
+		return fmt.Errorf("Clone(): %w", err)
+	}
+
+	logger.DebugContext(ctx, "cloned repo", "remote", g.remote, "branchOrTag", branchOrTag)
+
+	// Copy only the requested subdir to outDir.
+	if err := common.CopyRecursive(ctx, nil, &common.CopyParams{
+		DstRoot: outDir,
+		SrcRoot: subdirToCopy,
+		RFS:     &common.RealFS{},
+	}); err != nil {
+		return err //nolint:wrapcheck
+	}
+
+	return nil
+}
+
+// resolveBranchOrTag returns the latest release tag if branchOrTag is "latest", and otherwise
+// just returns the input branchOrTag after validating it. The return value always begins
+// with "v" (unless there's an error).
+func resolveBranchOrTag(ctx context.Context, t tagser, remote, branchOrTag string) (string, error) {
+	logger := logging.FromContext(ctx).With("logger", "resolveBranchOrTag")
+
+	if branchOrTag != "latest" {
```

Comment:
What's the behavior of empty `branchOrTag`? From the code, it seems still treated as "latest"? Just confirming it's intended.

Reply:
If branchOrTag is empty, then we'll enter this "if" block (because "" != "latest"). We'll then hit the ```fmt.Errorf(`the template source version %q must start with "v", like "v1.2.3"`, branchOrTag)``` a few lines below.

Reply:
But should we treat empty as "latest"?

Reply:
Perhaps so... @sethvargo do you have an opinion on whether `@latest` should be implicit if no version is given? It seems to work fine for `go get`.

Reply:
The upgrade behavior for Go modules is documented by confusing. I think getting latest makes sense.